### PR TITLE
Add scrolling to `_TextBox` widgets

### LINKED ISSUE #3445: Need to implement better text scrolling
### The issue:

This is to remind me that I need to:
* Allow mpris2 widget to keep scrolling
* Implement pixel scrolling

### Required:

- [X] I have searched past issues to see if this bug has already been reported.

Need to revist #2039 and #1927. 

I think pixel scrolling could be a mixin so, in theory, anything can be scrolled. We can think whether it's added to the base `_TextBox` class or whether there's a separate `ScrollingTextBox` class.

### PR DESCRIPTION
**Pull request by elParaguayo**
**Draft for now - testers wanted**

I've implemented scrolling in `_TextBox` widgets and would be keen for people to test.

The gif is a bit jerky (and loops in the wrong place) but you can see we're doing pixel scrolling.
out

Scrolling also works on vertical bars.

Things to do:
- [x] Testing (how?)
- [x] Replace scrolling in mpris2 widget with this
- [x] Fix bugs people find

Fixes #3445

### CODE REVIEW

File: libqtile/widget/base.py
Comment (m-col) [inline libqtile/widget/base.py:411]: It would help with clarity if all scroll-related settings had the same prefix, like the ones with `scroll_X`. Maybe `enabled_scrolling` can just be `scroll`, then `scroll_repeat`, `scroll_reset_on_update`, `scroll_hide_when_done` or something.

re the settings themselves, are these all settings that you think would be immediately useful (like in test widgets you may have configured), or are you predicting what might be useful? It's just I'm a bit skeptical about the usefulness of e.g. `repeat_scrolling` (wouldn't everybody want `True`?), `reset_scroll_on_update` (same). The role of `scroll_clear` and `hide_text_after_scroll` also seems similar - scroll once, then disappear, just that ones scrolls away rather than vanishes? How do these interact with `repeat_scrolling`, and do the constraints on how they can be used together suggest that the configuration options can be made more concise?

Lastly, is `scroll_width` not redundant with `width`, as only one can be used, but conceptually they mean something similar? Could `width` be used the way `scroll_width` is being used when `enable_scrolling = True`?

**Review by m-col — COMMENTED** [review]
Awesome, thanks for working on this! Looks cool. Gonna have to give it a proper out.

File: libqtile/widget/base.py
Comment (elParaguayo) [reply libqtile/widget/base.py:411]: Thanks. All good points. 

This is what I started with to get things working but I agree we can probably tidy up a bit.

I'm just playing with mpris2 widget but can refactor after that.

Comment (elParaguayo) [reply libqtile/widget/base.py:411]: `scroll_clear`: when `True` this keeps scrolling until all the text has scrolled out of the widget. When `False` the scrolling stops as soon as the end of the text is visible.
`hide_text_after_scroll`: This scrolls once and then hides the text.

Comment (elParaguayo) [reply libqtile/widget/base.py]: Made some changes. Haven't changed `width` yet. One reason for this is that it doesn't appear in `defaults` and I wanted to make it clearer but I definitely see the value in your approach too.

**Comment by elParaguayo** [comment]
I think I'll split this into 2 PRs to keep the mpris widget changes separate.

**Comment by elParaguayo** [comment]
> Could you make the default value for `scroll_repeat` True?

Yes.

**Comment by elParaguayo** [comment]
Added some tests. Hopefully good to get this one merged now.

File: libqtile/widget/base.py
Comment (m-col) [inline libqtile/widget/base.py:394]: "When True, you must set the widget's ``width``" - so if this won't work with flexible widgets. Is that a technical limitation, or does it not make sense conceptually?

Comment (elParaguayo) [reply libqtile/widget/base.py:394]: To scroll, you've got to have a fixed/max width, otherwise the widget won't scroll. This is why I split out `width` and `scroll_width` originally.

Comment (m-col) [reply libqtile/widget/base.py:394]: > To scroll, you've got to have a fixed/max width, otherwise the widget won't scroll.

But what's the reason for this limitation?

**Comment by m-col** [comment]
I pulled down the PR and tested it with the `Mpd2` widget, and got a configuration erorr complaining that `width` had been passed multiple times. I'm just debugging that now but I wonder what other widgets won't immediately work with this.

File: libqtile/widget/base.py
Comment (m-col) [inline libqtile/widget/base.py:497]:
```suggestion
            logger.warning("%s: You must specify a width when enabling scrolling.", self.name)
```

**Comment by m-col** [comment]
My error is because `width=bar.CALCULATED` is hardcoded into `ThreadPoolText`, which means this widget can't scroll currently. I suspect this isn't necessary though, maybe we can update `ThreadPoolText` to let it scroll?

File: libqtile/widget/base.py
Comment (elParaguayo) [reply libqtile/widget/base.py:394]: If the text box resizes to the width of the text then there's no point in scrolling. You scroll when the text is wider than the container.

File: libqtile/widget/base.py
Comment (elParaguayo) [reply libqtile/widget/base.py:497]: Yes. Need to remember this is not string formatting, this is logger formatting so it's ok to use `%s`!

File: libqtile/widget/base.py
Comment (m-col) [reply libqtile/widget/base.py:394]: Huh, oh yeah :laughing:

**Comment by elParaguayo** [comment]
> I pulled down the PR and tested it with the `Mpd2` widget, and got a configuration erorr complaining that `width` had been passed multiple times. I'm just debugging that now but I wonder what other widgets won't immediately work with this.

Ah. Interesting. I would hope the tests picked this up but I know `test_widget_init_configure` won't as it just ignores `ConfigErrors` (that's a separate issue that we may want to deal with...)

**Comment by elParaguayo** [comment]
> My error is because `width=bar.CALCULATED` is hardcoded into `ThreadPoolText`, which means this widget can't scroll currently. I suspect this isn't necessary though, maybe we can update `ThreadPoolText` to let it scroll?

We just need to remove `width=bar.CALCULATED`  from `ThreadPoolText`. It actually shouldn't be there as it would always give an error if anyone tried to set the width, regardless of the changes here.

**Comment by elParaguayo** [comment]
Ok. Thing these changes should fix that issue. Would be great if you can check too.

File: libqtile/widget/base.py
Comment (m-col) [inline libqtile/widget/base.py:735]: The width kwarg will take the value of the width key in the `config` dict already:
```py
class Widget:
    def __init__(self, width=10, **config):
        print("width", width, "config", config)


Widget()
Widget(width=0)
Widget(width=0, var=0)
```
prints:
```
width 10 config {}
width 0 config {}
width 0 config {'var': 0}
```
so i dont think we should need anything like 733-4 besides the default argument, that you have added

Comment (elParaguayo) [reply libqtile/widget/base.py:735]: Well, what you really need to show is a separate widget inheriting `Widget` without a `width` argument but I think you're still right.

Comment (elParaguayo) [reply libqtile/widget/base.py:735]:
```
>>> class Widget:
...   def __init__(self, width=10, **config):
...     print(width, config)
...
>>> class GoodChild(Widget):
...   def __init__(self, **config):
...     Widget.__init__(self, **config)
...
>>> class BadChild(Widget):
...   def __init__(self, **config):
...     Widget.__init__(self, width=10, **config)
...
>>> Widget(width=15)
15 {}
<__main__.Widget object at 0x7fd22a474c70>
>>> GoodChild(width=15)
15 {}
<__main__.GoodChild object at 0x7fd22a4741c0>
>>> BadChild(width=15)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in __init__
TypeError: __main__.Widget.__init__() got multiple values for keyword argument 'width'
>>>
```

Comment (m-col) [reply libqtile/widget/base.py]: Ah, gotcha, the duplicated `width=` is only acceptable if it's in the signature, not in any calling code, where the duplicate has to be resolved. Cool. So maybe this needs to do something like `config.setdefault("width", bar.CALCULATED)`

Comment (elParaguayo) [reply libqtile/widget/base.py]: Don't think we even need that. We just need to stop the child classes (`ThreadPoolText` etc) from setting a default value in their call to `_TextBox.__init__`. If it's just in config then `_TextBox` should handle it ok. That's the `GoodChild` example above.

Comment (elParaguayo) [reply libqtile/widget/base.py]: Some other widgets have this same issue e.g. `GroupBox`. While these shouldn't have fixed widths, I'm inclined to remove the code that causes the issue.

e.g.
https://github.com/qtile/qtile/blob/5c28cbb283c4397fb001951d6f12d0c5058988ae/libqtile/widget/groupbox.py#L47-L48

**Comment by elParaguayo** [comment]
> Playing around with `Mpris2` I discovered that it isn't getting cleaned up because of its scrolling timer. We have to make sure they are e.g. when config is reloaded

I've actually made some other changes to the Mpris2 widget off the back of this PR. It was originally part of the PR but I thought it may be better to separate it. Part of that change is to clear the display when the widget is finalized and remove timers so that may fix your issue.

Shall I add the commit back to this PR?

**Comment by elParaguayo** [comment]
When you say "always scrolls", do you mean that it repeats when it gets to the end? If so, that's because we set `scroll_repeat` to `True` by default (https://github.com/qtile/qtile/pull/3452#issuecomment-1090315419)

**Comment by m-col** [comment]
I meant that, based on its current (pre this PR) documentation, you cannot configure it to never scroll its text. From a look at the code, setting `scroll_interval` to 0 or None may disable it, but it isn't clear.

**Comment by elParaguayo** [comment]
Pushed the Mpris2 changes.

File: libqtile/widget/base.py
Comment (m-col) [inline libqtile/widget/base.py:537]:
```suggestion
        size = self.bar.size
```
Looks like `size` is only used in place so may as well use `self.bar.size` inline.

File: libqtile/widget/windowtabs.py
Comment (m-col) [inline libqtile/widget/windowtabs.py:57]:
```suggestion
        width = config.pop("width", bar.STRETCH)
```

File: libqtile/widget/base.py
Comment (m-col) [inline libqtile/widget/base.py:426]: It looks like these are only used by `self`, so could we make them private?

**Review by m-col — APPROVED** [review]
Brilliant as always! I've put in some tiny suggestions but otherwise LGTM when you're ready!

File: libqtile/widget/mpris2widget.py
Comment (m-col) [inline libqtile/widget/mpris2widget.py:149]: I think the original form would keep consistency with the rest of the logging. Actually I think flake8 will catch this!

File: libqtile/widget/windowtabs.py
Comment (elParaguayo) [reply libqtile/widget/windowtabs.py:57]: Nice. Didn't know that one.

File: libqtile/widget/mpris2widget.py
Comment (elParaguayo) [reply libqtile/widget/mpris2widget.py:149]: I'm confused. I thought we needed the `%s` style for logging.

Comment (m-col) [reply libqtile/widget/mpris2widget.py:149]: Yep we do. That's what it was before, and this change replaced it with a `"{}".format()` call.

Comment (elParaguayo) [reply libqtile/widget/mpris2widget.py:149]: Ah. Sorry. I thought this was your suggested change but it's actually because I copied some code from an old commit!

**Comment by elParaguayo** [comment]
We can now use this widget to listen to message from all compatible players. The current player's name is also retrieved but, currently, I do nothing with it, other than make it available via `info`.

We could expose it as a field (e.g. `{player}`) that can be added via the string formatting parameters but I left it out as this could be messy when we don't have the player name.

Or we could try to find the player's icon and show that (but that might cause some issue with the scrolling...)

**Comment by m-col** [comment]
Ideally the widget would behave like other text widgets where there is a dictionary of strings to use to for formatting the displayed text. From what I can tell the widget doesn't allow that customisation at all. I don't think any icon would be necessary (though certainly cool... maybe it would be easier to implement as a separate widget in the same module). Either way, probably out of scope for this PR!

**Comment by elParaguayo** [comment]
> Ideally the widget would behave like other text widgets where there is a dictionary of strings to use to for formatting the displayed text. From what I can tell the widget doesn't allow that customisation at all. I don't think any icon would be necessary (though certainly cool... maybe it would be easier to implement as a separate widget in the same module). Either way, probably out of scope for this PR!

Yeah - I did think about having a dictionary for the various metadata fields.

Will leave for another PR!

**Comment by m-col** [comment]
Looks amazing, thanks again! LGTM if you'd like to merge.

**Pull request merged**
